Lib/inspect.py

```diff
@@ -2416,20 +2416,41 @@ def _signature_from_function(cls, func, skip_bound_arg=True,
     s = getattr(func, "__text_signature__", None)
     if s:
         return _signature_fromstr(cls, func, s, skip_bound_arg)
-
+    return _signature_from_code(cls, func.__code__,
+                                       globals=globals,
+                                       locals=locals,
+                                       eval_str=eval_str,
+                                       is_duck_function=is_duck_function,
+                                       func=func)
+
+
+def _signature_from_code(cls,
+                         func_code,
+                         globals=None,
+                         locals=None,
+                         eval_str=False,
+                         is_duck_function=False,
+                         func=None):
+    """Private helper function to get signature for
+    code objects.
+    """
     Parameter = cls._parameter_cls
 
     # Parameter information.
-    func_code = func.__code__
     pos_count = func_code.co_argcount
     arg_names = func_code.co_varnames
     posonly_count = func_code.co_posonlyargcount
     positional = arg_names[:pos_count]
     keyword_only_count = func_code.co_kwonlyargcount
     keyword_only = arg_names[pos_count:pos_count + keyword_only_count]
-    annotations = get_annotations(func, globals=globals, locals=locals, eval_str=eval_str)
-    defaults = func.__defaults__
-    kwdefaults = func.__kwdefaults__
+    if func is not None:
+        annotations = get_annotations(func, globals=globals, locals=locals, eval_str=eval_str)
+        defaults = func.__defaults__
+        kwdefaults = func.__kwdefaults__
+    else:
+        annotations = {}
+        defaults = None
+        kwdefaults = None
```

Comment:
I prefer to use affirmative conditions where possible (`if func is None; else` rather than `if func is not None; else`) -- the double negatives make my brain go a little fuzzy otherwise 😄

```suggestion
    if func is None:
        annotations = {}
        defaults = None
        kwdefaults = None
    else:
        annotations = get_annotations(func, globals=globals, locals=locals, eval_str=eval_str)
        defaults = func.__defaults__
        kwdefaults = func.__kwdefaults__
```

Reply:
But `if param is not None` is extremely common, and the negated negative makes it a positive!
It’s fine if the code wants to address first the case where param is passed something (not None).

Reply:
I still prefer my suggestion, but I don't feel strongly on this one.

Reply:
Yeap, I prefer to keep this as-is, because `func` is `not None` most of the time.